### PR TITLE
promql: make lookback and matrix selections left-open and right-closed

### LINKED ISSUE #13213: promql: Make range selections left-open and right-closed
### Proposal

For historical reasons, a range selector selects a closed interval, i.e. samples perfectly coinciding with the boundaries of the range are included in the selection.

For various reasons, it is more consistent and more helpful in practice to make the selection "left open" and "right closed", i.e. a sample coinciding with the "left" boundary (the one further in the past) is excluded from the selection, while at the "right" boundary (the one further towards the future) stays inclusive.

For example, with samples perfectly spaced every 1m (very common in practice), a range over 5m might currently select 5 or 6 samples, depending on the exact alignment. In practice, it almost always selects 5 samples, because perfect alignment rarely happens in real-world scenarios. However, in test scenarios, perfect alignment happens easily, even without intending to do so. So test cases can easily represent a case that is very rare in practice. With the proposed change, the range will always select 5 samples in the case of perfectly spaced samples.

This is, however, a breaking change, although the impact is mostly academic. Therefore, we should implement this change with the upcoming 3.0.0 release.

### PR DESCRIPTION
**Pull request by KofClubs**
Fixes #13213

### CODE REVIEW

**Comment by KofClubs** [comment]
This Issue/PR is under discussion and is just an example.

File: promql/engine.go
Comment (beorn7) [inline promql/engine.go:1096]: You don't really need these two, do you? (In this PR, you only set them and never read them.)

**Review by beorn7 — COMMENTED** [review]
Thanks for tackling this. This _might_ be the only change needed. Best to see by adding test cases (or observing which existing tests change their result in which way), which should mostly affect the framework tests in https://github.com/prometheus/prometheus/tree/main/promql/testdata .

File: promql/engine.go
Comment (KofClubs) [reply promql/engine.go:1096]: Thanks, I shall remove them.

**Comment by KofClubs** [comment]
> Best to see by adding test cases (or observing which existing tests change their result in which way), which should mostly affect the framework tests in https://github.com/prometheus/prometheus/tree/main/promql/testdata .

I am working on unit tests. I ran all the unit tests and tracked 26 errors that shall be fixed:
```bash
go test github.com/prometheus/prometheus/...
```

**Comment by beorn7** [comment]
Sorry for the delay. I'm back from vacation and will try to get to this ASAP. If anyone qualified beats me to it, even better.

**Comment by beorn7** [comment]
In the meantime, @KofClubs could you fix the DCO? It should be easy. See the details of the failed DCO test.

**Comment by beorn7** [comment]
@vpranckaitis that's a good point. The lookback should be "left-open", too. Maybe it is already. We have to check.

**Comment by beorn7** [comment]
I still haven't found time for a detailed review. I will try later this week or early next week. My apologies.

**Comment by KofClubs** [comment]
> In the meantime, @KofClubs could you fix the DCO? It should be easy. See the details of the failed DCO test.

OK. I shall fixup all commits to one after review and sign it off.

Sorry, I've been extremely busy recently. I shall work on this PR now, to resolve conflicts firstly.

**Comment by KofClubs** [comment]
With conflicts resolved and commits signed off, I think this PR is ready to be reviewed:)

File: cmd/promtool/testdata/rules.yml
Comment (beorn7) [inline cmd/promtool/testdata/rules.yml:28]: It would actually be more interesting to leave this at `5m` and make sure that the result is one lower than before.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:743]: Which is not true anymore.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:744]: Maybe, in this case, do a `1m1s` range or something to still provoke the overflow.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:755]: The point here is that two large numbers average to 0, so here again, we should have a longer range, like `2m` or `1m1s`.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:758]: See above.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:760]: So yeah, in these tests, too, we want the same samples in the range, so the range duration has to be extended (rather than adjusting the result).

File: promql/promqltest/testdata/histograms.test
Comment (beorn7) [inline promql/promqltest/testdata/histograms.test:97]: Your editor has replaced the tab with spaces here. Let's keep things as they are (especially because the next line still has a tab), see also changes below that are only whitespace changes.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Looks good in general. A few subtleties in the comments.

I will look into the question of the lookback delta being inclusive ASAP.

**Comment by beorn7** [comment]
About the lookback delta:

There is even a test for it, see https://github.com/prometheus/prometheus/blob/806073ad6311bdccf0bc102a00fa7eda9fb563fb/promql/promqltest/testdata/staleness.test#L38-L51

Which means that the lookback is "inclusive" right now, but it needs to be "exclusive", so essentially, in that test, the `at 300s` should also yield an empty result, is does the `at 301s` now. This should be another simple change and should be included in this PR.

**Comment by beorn7** [comment]
And yet another question: How to merge this?

To merge this into main before the v3 release, we need to hide the behavior behind a feature flag. However, that doesn't play well with tests. So we could have the behavior changes in a separate PR, where they are behind a feature flag, and not tested at all. And then we have this PR with all the tests, which will go into a v3 branch, which doesn't exist yet, but maybe we should create one soon.

Another option would be to only launch this feature with v3 and don't have it behind a feature flag in v2 at all.

WDYT?

File: promql/promqltest/testdata/functions.test
Comment (KofClubs) [reply promql/promqltest/testdata/functions.test]: Get it. I changed evaluation range to `2m`.

File: promql/promqltest/testdata/histograms.test
Comment (KofClubs) [reply promql/promqltest/testdata/histograms.test]: My bad. Corrected.

File: cmd/promtool/testdata/rules.yml
Comment (KofClubs) [reply cmd/promtool/testdata/rules.yml]: Seem not to change any behavior, so I undo this change.

File: promql/promqltest/testdata/functions.test
Comment (KofClubs) [reply promql/promqltest/testdata/functions.test]: Deleted.

File: promql/promqltest/testdata/functions.test
Comment (KofClubs) [reply promql/promqltest/testdata/functions.test:743]: `metric8` is declared as:
```
load 10s
  metric8 9.988465674311579e+307 9.988465674311579e+307
```
So `eval instant at 1m count_over_time(metric8[1m1ms])` shall be 2.

**Comment by KofClubs** [comment]
> About the lookback delta:

Get it. I think there are some aspects worth exploring.

#### Motivation
Apart from making behavior consistent, are there any other motivations for making lookback into `(now-5m, now]`?
It seems that the opening and closing of the left bound (`now-5m`) does not affect the correctness of the lookback behavior.

#### From Storage Perspective
I don't know if there is any bias in my understanding. If lookback becomes left-open and right-closed, then from the perspective of storage (which can be Prometheus itself, or another TSDB replaced by users), when retrieving data, the time interval shall become left-open and right-closed: `(start, end]`. Does Prometheus engine need to ask storage to ignore samples falling on `start`?

If there are any misunderstandings above, please feel free to point them out:)

**Comment by KofClubs** [comment]
> WDYT?

Both OK. While, I think this change has a large impact. Considering the change in user habits and behaviors, adding a flag seems to be necessary.
I think left-open and right-closed selection shall be enabled by default in Prometheus 3.0. Unless the user turns it on, like `--matrix-selector-version=2`, left-closed and right-closed selection shall never work. And no unit tests for v2-like behaviors shall be implemented in v3.

**Comment by beorn7** [comment]
WRT lookback delta: I think the one and only argument is the one made by @vpranckaitis already. "Correct" is whatever behavior we define as such.

The implementation details in the storage shouldn't matter much. The easiest way for now is probably to simple reduce the timestamp by 1ms. And it should be included in this PR to keep things consistent.

**Comment by beorn7** [comment]
About flags and configurability:

I think it is uncontroversial that v3 should only support the new behavior (the one implemented in this PR), with no way of bringing back the old behavior. So far, everyone that thought about it has agreed that the "left closed" behavior was a mistake and should have never happened in the first place.

The only question is how to deal with it in v2. And I see two feasible options:
1. Do a hard switch with v3, so v2 always has the old behavior and v3 always has the new behavior. Fullstop.
2. Make v2 configurable with a feature flag, so people can try out the new behavior without migrating to v3.

The problem with (2) is that it is hard to cover in tests (unless we add feature flag configuration to the testing framework), but that's maybe not a big problem because it will be tested in the v3 branch. On the other hand, the change in behavior is almost irrelevant in practice and mostly matters it tests (where you have perfectly aligned artificial data), so maybe a feature flag isn't even necessary and (1) is the way to go.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [reply promql/promqltest/testdata/functions.test:743]: We have a misunderstanding here. The purpose of this test is to create a `Inf` result. so the comment above should stay, and the test expression should be `eval instant at 1m sum_over_time(metric8[2m])/count_over_time(metric8[2m])` (or similar).

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [reply promql/promqltest/testdata/functions.test:759]: And again, the purpose of this test is to cover all four values, so the test expression should be something like `eval instant at 45s sum_over_time(metric[1m])`, and the result should be unchanged.

**Comment by beorn7** [comment]
I unresolved two conversation above where you went into the opposite direction than intended.

File: promql/promqltest/testdata/functions.test
Comment (KofClubs) [reply promql/promqltest/testdata/functions.test:743]: Get it. Resolved.

File: promql/promqltest/testdata/functions.test
Comment (KofClubs) [reply promql/promqltest/testdata/functions.test:759]: Resolved. I added such cases:
```
eval instant at 45s avg_over_time(metric10[1m])
  {} 0

eval instant at 1m avg_over_time(metric10[2m])
  {} 0

eval instant at 45s sum_over_time(metric10[1m])/count_over_time(metric10[1m])
  {} 0

eval instant at 1m sum_over_time(metric10[2m])/count_over_time(metric10[2m])
  {} 0
```
Both `metric10[1m]` at 45s and `metric10[2m]` at 1m shall evaluate 2 samples, to make them `0/2`.

**Comment by KofClubs** [comment]
> I unresolved two conversation above where you went into the opposite direction than intended.

Thanks. Both resolved. Review them please:)

**Comment by KofClubs** [comment]
> (1) is the way to go.

Merging this update to v3 directly sounds good:)

As there are significant functional differences between v2 and v3, are there plans to continue maintaining v2 after v3 is released? When shall the life cycle of v2 end?

Sorry for not knowing about the planning of Prometheus 3.0 fully, so I would like to ask these questions sincerely:)

**Comment by KofClubs** [comment]
Several unit tests to be changed.

I shall continue to work on them.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:965]: Same pattern here (essentially, if there is less than a minute of time series data, and the test query is `at 1m ... [1m]`, it probably wants to touch all the samples, but with the new range definition, we now exclude the 1st sample, so we need a longer range. Same for line 958 above. It's only by accident that the results don't change. In both cases, use a longer range, e.g. 2m.

File: promql/promqltest/testdata/subquery.test
Comment (beorn7) [inline promql/promqltest/testdata/subquery.test:105]: The test intends to create 2.366666667 as a result, as per the formula above. So we need to change the ranges in the subquery accordingly. I suggest to use `rate(metric[1m5s:10s])`

File: promql/promqltest/testdata/subquery.test
Comment (beorn7) [inline promql/promqltest/testdata/subquery.test:113]: The tests intends to touch two samples, so the range has to be changed. I propose `min_over_time(metric[15s:10s])`.

File: promql/promqltest/testdata/subquery.test
Comment (beorn7) [inline promql/promqltest/testdata/subquery.test:57]: While I think it's fine in this case to leave the expression unchanged (and change the expected result, as you did), you need to update the explanation in lines 54 to 56. It's just not true anymore with the new range definition.

File: promql/promqltest/testdata/subquery.test
Comment (beorn7) [inline promql/promqltest/testdata/subquery.test:64]: Interestingly, this test previously proved the shortcomings of the old definition: With the perfect alignment, you got a different result (394) from an imperfect alignment (297, see below). With the new definition, the result is always 297. So your change to the expected result in line 61 is fine, but the explanations around it needed to be updated.

**Review by beorn7 — CHANGES_REQUESTED** [review]
A few more instances where we either need a new explanation, or we should change the query rather than the expected result.

File: promql/engine.go
Comment (beorn7) [inline promql/engine.go:503]: Maybe let's do it differently (and in a cleaner way) after all. I'll suggest something in a separate comment.

**Comment by beorn7** [comment]
I pushed a commit to your branch to show my idea about adjusting the lookback delta. I think that is cleaner. Please double-check if that works, and feel free to squash my commit into yours.

**Comment by beorn7** [comment]
My commit breaks all kind of tests. I'll try to find out why and update accordingly. Stay tuned.

**Comment by beorn7** [comment]
> As there are significant functional differences between v2 and v3, are there plans to continue maintaining v2 after v3 is released? When shall the life cycle of v2 end?

I expect RCs for 3.0 for a while, and once the final 3.0.0 is out, the LTS of v2 will be maintained for a while as usual, but probably only with bugfixes. But that's just my speculation. Details need to be fleshed out by the v3 coordinators.

**Comment by beorn7** [comment]
I think I got all tests passing. Let's see if it goes green.

If it does, please still double check my lookback delta changes.

Then you only need to address my remaining comments about the tests above, and we are done.

**Comment by KofClubs** [comment]
@beorn7 Thanks for detailed code review.

I understand your comments on `promql/promqltest/testdata/functions.test` and made eval range fully cover the input samples ​​set intentionally.

I also updated the comments involved in the changes of `promql/promqltest/testdata/subquery.test`.

**Comment by KofClubs** [comment]
I noticed that `promql/engine_test.go` is constantly changing, especially at the code blocks commented as `// TODO(.*): Integrate .* into the PromQL testing framework`.

So how shall this PR be merged?

**Comment by beorn7** [comment]
> I noticed that `promql/engine_test.go` is constantly be changed, especially at the code blocks commented as `// TODO(.*): Integrate .* into the PromQL testing framework`.
> 
> So how shall this PR be merged?

Given that the changes we need for this overall are sprinkled over many places, I would prefer if we didn't offer the changed behavior as a feature flag in v2.

So we need to wait for a v3 release branch to show up (I'm already talking to the v3 coordinators). Then this PR has to be rebased on top of that branch. Then we merge it into the v3 release branch.

And yes, then the branch has to be kept up to date with changes in the main branch. That will be a bit tedious, but I guess that's unavoidable in this kind of situation. (And once this PR is merged, it's not your job anymore to keep the v3 branch up to date. ;)

**Comment by beorn7** [comment]
I'll do the final code level review of this PR ASAP.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:11]: It's good to test that the 5m range yields an empty result, but let's also add a new test with 10m range, to show that it has the same result as the previous 5m test.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:18]: In a similar spirit, let's add a 32m range to demonstrate the two counter resets for `/foo`.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:34]: Let's add a 6m test here to demonstrate this same result as the old 5m test.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:60]: We want to see the two changes here. Let's extend the range to 20m.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:74]: The test was designed to yield a round result. Let's extend the range to 55m here to keep it that way.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:112]: Here we would also like to see a round result. Does it work if you extend the range to 55m?

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:122]: The point of this test is to yield 0, not an empty result, so you need to extend the range here. 6m should work.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:245]: The comments explain the expected results here. We either have to change both comments and results, or change the ranges so that the comments still apply.

I recommend the latter. Using a 55m range should do the trick for each of those.

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [inline promql/promqltest/testdata/functions.test:677]: The tests intend to use all the samples stated. Let's just shift the eval time a bit to keep it that way, , i.e. `at 1m` → `at 55s`.

File: promql/promqltest/testdata/operators.test
Comment (beorn7) [inline promql/promqltest/testdata/operators.test:117]: Tab vs. spaces issue here. (Next line has a leading tab, this one has four leading spaces.)

**Review by beorn7 — CHANGES_REQUESTED** [review]
I found a few more tests where we have to change the range or the eval time rather than the result. Sorry for not spotting them earlier. These are a lot of tests.

Also some other issues caused by the resolution of the merge conflicts.

Finally, there is one thing left: We have to update the documentation. I'll ad some pointers later.

**Comment by beorn7** [comment]
In different news: The v3 coordinators will discuss cutting a release branch next week.

**Comment by beorn7** [comment]
Documentation updates needed (probably not a complete list, just the locations I could find at a glance):

Needs clarification that the lookback period excludes samples that are precisely 5m earlier that the evaluation time: https://github.com/prometheus/prometheus/blob/2544066169b9f9cd3cf9693a71d9da2fb215856a/docs/querying/basics.md?plain=1#L338-L339 

Closed range interval description needs update: https://github.com/prometheus/prometheus/blob/2544066169b9f9cd3cf9693a71d9da2fb215856a/docs/querying/basics.md?plain=1#L192-L194

**Comment by KofClubs** [comment]
Some unit tests were significantly broken, because this change made extrapolation much more common.

Docs has been updated:)

File: promql/engine_test.go
Comment (beorn7) [reply promql/engine_test.go:3493]: This is still unresolved.

File: promql/engine_test.go
Comment (beorn7) [inline promql/engine_test.go:3493]:
```suggestion
				offset := int64(len(c.histograms) - 1)
				newTs := ts + offset*int64(time.Minute/time.Millisecond)
```

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [reply promql/promqltest/testdata/functions.test:89]: OK, I played with it. I still think we should change the test so that it yields round numbers, but that requires some more changes. Here is my suggestion:

```Go
# Tests for increase().
load 5m
	http_requests{path="/foo"}	0+10x11
	http_requests{path="/bar"}	0+10x5 0+10x5
	http_requests{path="/dings"}   10+10x11
	http_requests{path="/bumms"}    1+10x11

eval instant at 55m increase(http_requests[55m])
	{path="/foo"}   110
	{path="/bar"}    99
	{path="/dings"} 110
	{path="/bumms"} 110
```

File: promql/promqltest/testdata/functions.test
Comment (beorn7) [reply promql/promqltest/testdata/functions.test:127]: Again, we need to change a few more things. Here my suggestion:

```Go
# Tests for rate().
load 5m
	testcounter_reset_middle	10+10x4 0+10x5
	testcounter_reset_end    	10+10x9 0 10

# Counter resets at in the middle of range are handled correctly by rate().
eval instant at 50m rate(testcounter_reset_middle[55m])
	{} 0.03
```

**Review by beorn7 — CHANGES_REQUESTED** [review]
There are three unresolved comments, with my now more detailed suggestions how to resolve them.

File: docs/querying/basics.md
Comment (beorn7) [inline docs/querying/basics.md:194]:
```suggestion
excluded from the selection, while samples coinciding with the right boundary of the range are
```

File: docs/querying/basics.md
Comment (beorn7) [inline docs/querying/basics.md:198]: Maybe let's word this in a more straightforward way:

```suggestion
In this example, we select all the values recorded less than 5m ago for all time series
```

File: docs/querying/basics.md
Comment (beorn7) [inline docs/querying/basics.md:340]: I'm not sure if "left-open" will be understood in this context. How about the following:

```suggestion
series. It does so by taking the newest sample that is less than the lookback period ago.
```

**Review by beorn7 — CHANGES_REQUESTED** [review]
A few more comments about the documentation.

**Comment by beorn7** [comment]
Thank you very much. We are almost done here. Just the few comments above, all with suggestions what to change.

I have proposed to the v3 coordinators to start a release branch soon. Let's see when it happens.

**Comment by KofClubs** [comment]
@beorn7 All resolved. Thank you!

**Comment by beorn7** [comment]
Looks good. Thank you very much.

(I'm not formally approving this PR right now, because we don't want to merge it into main. We need the v3 release branch to exist before merging.)

**Comment by beorn7** [comment]
We have a v3 branch: https://github.com/prometheus/prometheus/tree/release-3.0

I have already changed the merge target for this PR.

However, there are merge conflicts again (have been there before I changed the merge target).

@KofClubs could you rebase on top of the current state of the new release-3.0 branch and address the merge conflicts?

**Comment by KofClubs** [comment]
@beorn7 Done. Thank you:)

**Review by beorn7 — APPROVED** [review]

**Comment by beorn7** [comment]
Great stuff. Will merge on green.

**Pull request merged**
